The index mustn't be greater than the length of the string, 18, but it was 255.

[2018-11-20 12:05:15{UTC}] [ERROR] [http-apr-8080-exec-43] [c.q.z.s.a.ServiceExceptionAspect] [Got exception when calling [String com.qaprosoft.zafira.services.services.application.EmailService.sendEmail(IEmailMessage, String[])]]
com.qaprosoft.zafira.services.exceptions.ServiceException: ?substring(...) argument #2 had invalid value: The index mustn't be greater than the length of the string, 18, but it was 255.

----
FTL stack trace ("~" means nesting-related):
        - Failed at: ${testRun.comments?trim?substring(0, ...  [in template "test_run_results.ftl" at line 93, column 25]
----
        at com.qaprosoft.zafira.services.util.FreemarkerUtil.getFreeMarkerTemplateContent(FreemarkerUtil.java:44) ~[zafira-services-2.0.jar:na]
        at com.qaprosoft.zafira.services.services.application.EmailService.sendEmail(EmailService.java:70) ~[zafira-services-2.0.jar:na]
        at com.qaprosoft.zafira.services.services.application.TestRunService.sendTestRunResultsNotification(TestRunService.java:564) [zafira-services-2.0.jar:na]
        at com.qaprosoft.zafira.services.services.application.TestRunService.sendTestRunResultsEmail(TestRunService.java:544) [zafira-services-2.0.jar:na]
        at com.qaprosoft.zafira.services.services.application.TestRunService$$FastClassBySpringCGLIB$$bd22647b.invoke(<generated>) [zafira-services-2.0.jar:na]
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204) [spring-core-4.3.20.RELEASE.jar:4.3.20.RELEASE]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:736) [spring-aop-4.3.20.RELEASE.jar:4.3.20.RELEASE]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157) [spring-aop-4.3.20.RELEASE.jar:4.3.20.RELEASE]
        at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99) [spring-tx-4.3.20.RELEASE.jar:4.3.20.RELEASE]
        at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:282) [spring-tx-4.3.20.RELEASE.jar:4.3.20.RELEASE]
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96) [spring-tx-4.3.20.RELEASE.jar:4.3.20.RELEASE]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) [spring-aop-4.3.20.RELEASE.jar:4.3.20.RELEASE]

--- a/sources/zafira-models/src/main/java/com/qaprosoft/zafira/models/db/TestRun.java
+++ b/sources/zafira-models/src/main/java/com/qaprosoft/zafira/models/db/TestRun.java
@@ -169,8 +169,12 @@ public class TestRun extends AbstractEntity
 		}
 		String appVersion = argumentIsPresent("app_version")? this.configuration.get("app_version") + " - ": "";
 		String platformInfo = buildPlatformInfo();
-		return String.format(NAME, appVersion, testSuite.getName(), testSuite.getFileName(),
+		String testRunName = String.format(NAME, appVersion, testSuite.getName(), testSuite.getFileName(),
 				this.configuration.get("env"), platformInfo).trim();
+		if(env == null){
+			testRunName = testRunName.split("on")[0];
+		}
+		return testRunName;
 	}
 
 	private boolean argumentIsPresent(String arg, String... ignoreValues) {

--- a/sources/zafira-models/src/main/java/com/qaprosoft/zafira/models/db/TestRun.java
+++ b/sources/zafira-models/src/main/java/com/qaprosoft/zafira/models/db/TestRun.java
@@ -28,6 +28,7 @@ import org.joda.time.Seconds;
 import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.annotation.JsonInclude;
 import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import org.springframework.util.StringUtils;
 
 @JsonInclude(Include.NON_NULL)
 public class TestRun extends AbstractEntity
@@ -171,7 +172,7 @@ public class TestRun extends AbstractEntity
 		String platformInfo = buildPlatformInfo();
 		String testRunName = String.format(NAME, appVersion, testSuite.getName(), testSuite.getFileName(),
 				this.configuration.get("env"), platformInfo).trim();
-		if(env == null){
+		if(StringUtils.isEmpty(this.configuration.get("env"))){
 			testRunName = testRunName.split("on")[0];
 		}
 		return testRunName;

--- a/sources/zafira-services/src/main/java/com/qaprosoft/zafira/services/services/application/TestRunService.java
+++ b/sources/zafira-services/src/main/java/com/qaprosoft/zafira/services/services/application/TestRunService.java
@@ -16,11 +16,7 @@
 package com.qaprosoft.zafira.services.services.application;
 
 import static com.qaprosoft.zafira.models.db.Setting.SettingType.JIRA_URL;
-import static com.qaprosoft.zafira.models.db.Status.ABORTED;
-import static com.qaprosoft.zafira.models.db.Status.FAILED;
-import static com.qaprosoft.zafira.models.db.Status.IN_PROGRESS;
-import static com.qaprosoft.zafira.models.db.Status.PASSED;
-import static com.qaprosoft.zafira.models.db.Status.SKIPPED;
+import static com.qaprosoft.zafira.models.db.Status.*;
 import static com.qaprosoft.zafira.services.util.DateFormatter.actualizeSearchCriteriaDate;
 
 import java.io.ByteArrayInputStream;
@@ -82,6 +78,14 @@ public class TestRunService
 {
 	private static final Logger LOGGER = LoggerFactory.getLogger(TestRunService.class);
 	public static final String DEFAULT_PROJECT = "UNKNOWN";
+
+	public enum FailureCause {
+		UNRECOGNIZED_FAILURE,
+		COMPILATION_FAILURE,
+		TIMED_OUT,
+		BUILD_FAILURE,
+		ABORTED
+	}
 
 	@Autowired
 	private URLResolver urlResolver;
@@ -411,7 +415,7 @@ public class TestRunService
 			{
 				for(Test test : tests)
 				{
-					if(IN_PROGRESS.equals(test.getStatus()))
+					if(IN_PROGRESS.equals(test.getStatus()) || QUEUED.equals(test.getStatus()) && isBuildFailure(testRun.getComments()))
 					{
 						testService.abortTest(test, abortCause);
 					}
@@ -605,6 +609,17 @@ public class TestRunService
 		double rate = (double) testRun.getPassed() / (double) total;
 		return total > 0 ? (new BigDecimal(rate).setScale(2, RoundingMode.HALF_UP).multiply(new BigDecimal(100))).intValue() : 0;
 	}
+
+    public boolean isBuildFailure(String comments)
+    {
+        boolean failure = false;
+        if(StringUtils.isNotEmpty(comments)){
+            if(comments.contains(FailureCause.BUILD_FAILURE.name()) || comments.contains(FailureCause.COMPILATION_FAILURE.name())){
+                failure = true;
+            }
+        }
+        return failure;
+    }
 
 	@Transactional
 	public TestRun addComment(long id, String comment) throws ServiceException

--- a/sources/zafira-services/src/main/java/com/qaprosoft/zafira/services/services/application/TestRunService.java
+++ b/sources/zafira-services/src/main/java/com/qaprosoft/zafira/services/services/application/TestRunService.java
@@ -411,7 +411,7 @@ public class TestRunService
 	{
 		if(testRun != null){
 			List<Test> tests = testService.getTestsByTestRunId(testRun.getId());
-			if(IN_PROGRESS.equals(testRun.getStatus()) || QUEUED.equals(testRun.getStatus()) && isBuildFailure(testRun.getComments()))
+			if(IN_PROGRESS.equals(testRun.getStatus()) || QUEUED.equals(testRun.getStatus()) && isBuildFailure(abortCause))
 			{
 				for(Test test : tests)
 				{

--- a/sources/zafira-services/src/main/java/com/qaprosoft/zafira/services/services/application/TestRunService.java
+++ b/sources/zafira-services/src/main/java/com/qaprosoft/zafira/services/services/application/TestRunService.java
@@ -411,11 +411,11 @@ public class TestRunService
 	{
 		if(testRun != null){
 			List<Test> tests = testService.getTestsByTestRunId(testRun.getId());
-			if(IN_PROGRESS.equals(testRun.getStatus()))
+			if(IN_PROGRESS.equals(testRun.getStatus()) || QUEUED.equals(testRun.getStatus()) && isBuildFailure(testRun.getComments()))
 			{
 				for(Test test : tests)
 				{
-					if(IN_PROGRESS.equals(test.getStatus()) || QUEUED.equals(test.getStatus()) && isBuildFailure(testRun.getComments()))
+					if(IN_PROGRESS.equals(test.getStatus()) || QUEUED.equals(test.getStatus()))
 					{
 						testService.abortTest(test, abortCause);
 					}

--- a/sources/zafira-services/src/main/java/com/qaprosoft/zafira/services/services/application/TestRunService.java
+++ b/sources/zafira-services/src/main/java/com/qaprosoft/zafira/services/services/application/TestRunService.java
@@ -634,13 +634,6 @@ public class TestRunService
 		return testRunMapper.getTestRunByIdFull(id);
 	}
 
-	@Transactional
-	public void updateComment(TestRun testRun, String comment) throws ServiceException
-	{
-		testRun.setComments(comment);
-		updateTestRun(testRun);
-	}
-
 	@Transactional(readOnly = true)
 	@Cacheable(value = "environments", key = "T(com.qaprosoft.zafira.dbaccess.utils.TenancyContext).tenantName + ':' + #result", condition = "#result != null && #result.size() != 0")
 	public List<String> getEnvironments() throws ServiceException

--- a/sources/zafira-services/src/main/resources/templates/test_run_results.ftl
+++ b/sources/zafira-services/src/main/resources/templates/test_run_results.ftl
@@ -90,7 +90,7 @@
                 <td>Comments:</td>
                 <td>
                     <pre style="white-space: pre-line;">
-                        ${testRun.comments?trim[0..<255]}
+                        ${testRun.comments?trim[0..*255]}
                     </pre>
                 </td>
             </tr>

--- a/sources/zafira-services/src/main/resources/templates/test_run_results.ftl
+++ b/sources/zafira-services/src/main/resources/templates/test_run_results.ftl
@@ -90,7 +90,7 @@
                 <td>Comments:</td>
                 <td>
                     <pre style="white-space: pre-line;">
-                        ${testRun.comments?trim?substring(0, 255)}
+                        ${testRun.comments?trim[0..<255]}
                     </pre>
                 </td>
             </tr>

--- a/sources/zafira-ws/src/main/java/com/qaprosoft/zafira/ws/controller/application/TagsAPIController.java
+++ b/sources/zafira-ws/src/main/java/com/qaprosoft/zafira/ws/controller/application/TagsAPIController.java
@@ -28,6 +28,7 @@ import io.swagger.annotations.Api;
 import io.swagger.annotations.ApiImplicitParam;
 import io.swagger.annotations.ApiImplicitParams;
 import io.swagger.annotations.ApiOperation;
+import org.apache.commons.lang.StringUtils;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.MediaType;
@@ -61,16 +62,26 @@ public class TagsAPIController extends AbstractController
             integration.setCreatedAfter(testRun.getCreatedAt().getTime());
             Configuration configuration = testRunService.readConfiguration(testRun.getConfigXML());
             integration.setTestRunName(testRun.getName(configuration));
-            if(integrationTag == IntegrationTag.TESTRAIL_TESTCASE_UUID){
-                configuration.getArg().forEach(arg -> {
-                    if(arg.getKey().contains("testrail_assignee")){
-                        integration.getCustomParams().put("assignee", arg.getValue());
-                    } else if (arg.getKey().contains("testrail_milestone")){
-                        integration.getCustomParams().put("milestone", arg.getValue());
-                    } else if (arg.getKey().contains("testrail_run_name")){
-                        integration.setTestRunName(arg.getValue());
-                    }
-                });
+            switch (integrationTag){
+                case TESTRAIL_TESTCASE_UUID:
+                    configuration.getArg().forEach(arg -> {
+                        if(arg.getKey().contains("testrail_assignee")){
+                            integration.getCustomParams().put("assignee", arg.getValue());
+                        } else if (arg.getKey().contains("testrail_milestone")){
+                            integration.getCustomParams().put("milestone", arg.getValue());
+                        } else if (arg.getKey().contains("testrail_run_name") && !StringUtils.isEmpty(arg.getValue())){
+                            integration.setTestRunName(arg.getValue());
+                        }
+                    });
+                    break;
+                case QTEST_TESTCASE_UUID:
+                    configuration.getArg().forEach(arg -> {
+                        if(arg.getKey().contains("qtest_cycle_name")){
+                            integration.getCustomParams().put("cycle_name", arg.getValue());
+                        } else if (arg.getKey().contains("qtest_suite_name") && !StringUtils.isEmpty(arg.getValue())){
+                            integration.setTestRunName(arg.getValue());
+                        }
+                    });
             }
         }
         return integration;

--- a/sources/zafira-ws/src/main/java/com/qaprosoft/zafira/ws/controller/application/TestRunsAPIController.java
+++ b/sources/zafira-ws/src/main/java/com/qaprosoft/zafira/ws/controller/application/TestRunsAPIController.java
@@ -438,13 +438,9 @@ public class TestRunsAPIController extends AbstractController {
 		if (testRun == null) {
 			throw new TestRunNotFoundException();
 		}
-		if(!StringUtils.isEmpty(testRun.getComments())) {
-			if(rerunFailures){
-				testRunService.updateComment(testRun, "rebuild failures run");
-			} else {
-				testRunService.updateComment(testRun, null);
-			}
-		}
+        testRun.setComments(null);
+        testRun.setReviewed(false);
+        testRunService.updateTestRun(testRun);
 		if (!jenkinsService.rerunJob(testRun.getJob(), testRun.getBuildNumber(), rerunFailures)) {
 			throw new UnableToRebuildCIJobException();
 		}

--- a/sources/zafira-ws/src/main/java/com/qaprosoft/zafira/ws/controller/application/TestRunsAPIController.java
+++ b/sources/zafira-ws/src/main/java/com/qaprosoft/zafira/ws/controller/application/TestRunsAPIController.java
@@ -438,9 +438,9 @@ public class TestRunsAPIController extends AbstractController {
 		if (testRun == null) {
 			throw new TestRunNotFoundException();
 		}
-        testRun.setComments(null);
-        testRun.setReviewed(false);
-        testRunService.updateTestRun(testRun);
+                testRun.setComments(null);
+                testRun.setReviewed(false);
+                testRunService.updateTestRun(testRun);
 		if (!jenkinsService.rerunJob(testRun.getJob(), testRun.getBuildNumber(), rerunFailures)) {
 			throw new UnableToRebuildCIJobException();
 		}

